src/assets/CurrencyRateController.test.ts

```diff
@@ -34,19 +33,77 @@ const getStubbedDate = () => {
   return new Date('2019-04-07T10:20:30Z').getTime();
 };
 
+/**
+ * Setup a test CurrencyRateController instance.
+ *
+ * @param options - the options object for setupping up the CurrencyRateController for testing.
+ * @param options.state - the initial state for the CurrencyRateController.
+ * @param options.state.nativeCurrency - the initial nativeCurrency at time of instantiation.
+ * @param options.state.currentCurrency - the initial currentCurrency at time of instantiation.
+ * @param options.interval - the interval at which to poll for updated exchange rates.
+ * @param options.conversionRate - the conversionRate to set for the fetchExchangeRateStub.
+ * @param options.includeUsdRate - a boolean indicating whether or not to fetch/return a usdConversion rate.
+ * @param options.usdConversionRate - the usdConversionRate to set for the fetchExchangeRateStub.
+ * @param options.fetchExchangeRateStub - a full stub of the fetchExchangeRate method used by the Currency rate.
+ * @returns an object containing the setup CurrencyRateController and the fetchExchangeRateStub.
+ */
+function setupController(
+  {
+    state,
+    interval,
+    conversionRate,
+    includeUsdRate,
+    usdConversionRate,
+    fetchExchangeRateStub,
+  }: {
+    conversionRate?: number;
+    usdConversionRate?: number;
+    interval?: number;
+    state?: { nativeCurrency?: string; currentCurrency?: string };
+    includeUsdRate?: boolean;
+    fetchExchangeRateStub?:
+      | ((
+          currency: string,
+          nativeCurrency: string,
+          includeUSDRate?: boolean | undefined,
+        ) => Promise<{ conversionRate: number; usdConversionRate: number }>)
+      | undefined;
+  } = {
+    conversionRate: 1,
+    usdConversionRate: 2,
+    interval: 100,
+    state: { nativeCurrency: 'ETH' },
+    includeUsdRate: false,
+  },
+) {
+  let fetchExchangeRate;
+  if (fetchExchangeRateStub) {
+    fetchExchangeRate = fetchExchangeRateStub;
+  } else if (conversionRate || usdConversionRate) {
+    fetchExchangeRate = jest
+      .fn()
+      .mockImplementation(() =>
+        Promise.resolve({ conversionRate, usdConversionRate }),
+      );
+  }
+  const controller = new CurrencyRateController({
+    fetchExchangeRate,
+    messenger: getRestrictedMessenger(),
+    interval,
+    state,
+    includeUsdRate,
+  });
+
+  return { controller, fetchExchangeRate };
```

Comment:
So it looks like the CurrencyRateController has a `destroy` method. What do you think about having that get called automatically? I don't have an example handy right now but basically instead of `setupController` you could have a `withController` function that calls a function and passes a new instance of the controller, then makes sure to call `destroy` on the controller. Something like this:

``` typescript
type WithControllerOptions = {
  // you have this above already
}

type WithControllerCallback = ({
  controller: CurrencyRateController
  fetchExchangeRate: (
    currency: string,
    nativeCurrency: string,
    includeUSDRate?: boolean | undefined,
  ) => Promise<{ conversionRate: number; usdConversionRate: number }>
}) => void

function withController(...args: [WithControllerOptions] | [WithControllerOptions, WithControllerCallback]) {
  const [options, callback] = args.length === 2 ? args : [{}, args[0]];
  const {
    state,
    interval,
    conversionRate,
    includeUsdRate,
    usdConversionRate,
    fetchExchangeRateStub,
  } = options;
  // ... instantiate the controller, etc ...
  try {
    callback({ controller, fetchExchangeRate });
  } finally {
    controller.destroy();
  }
}
```

Now in tests you'd write:

``` typescript
it('should do whatever', () => {
  withController(
    ({ controller }) => {
      // do whatever
    })
  )
})
```

or:

``` typescript
it('should do whatever', () => {
  withController(
    { state: existingState },
    ({ controller, fetchExchangeRate }) => {
      // do whatever
    })
  )
})
```

instead of e.g.

``` typescript
it('should do whatever', () => {
  const { controller, fetchExchangeRate } = setupController();
  // do whatever
})
```

What do you think about this change?

Reply:
@mcmire will look into this! Focusing on trying to cut a new controllers patch release first